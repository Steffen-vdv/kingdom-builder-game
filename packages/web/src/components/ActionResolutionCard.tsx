import React from 'react';
import type { ActionResolution } from '../state/useActionResolution';
import {
	CARD_BASE_CLASS,
	CARD_BODY_TEXT_CLASS,
	CARD_LABEL_CLASS,
	CARD_META_TEXT_CLASS,
	CARD_TITLE_TEXT_CLASS,
	CONTINUE_BUTTON_CLASS,
	joinClasses,
} from './common/cardStyles';

interface ActionResolutionCardProps {
	title?: string;
	resolution: ActionResolution;
	onContinue: () => void;
}

function ActionResolutionCard({
	title,
	resolution,
	onContinue,
}: ActionResolutionCardProps) {
	const playerLabel = resolution.player?.name ?? resolution.player?.id ?? null;
	const playerName = playerLabel ?? 'Unknown player';
	const containerClass = `${CARD_BASE_CLASS} pointer-events-auto`;
	const headerTitle = title ?? 'Action resolution';
	const actionName = resolution.action?.name?.trim();
	const actionIcon = resolution.action?.icon?.trim();
	const summaryItems = resolution.summaries.filter((item): item is string =>
		Boolean(item?.trim()),
	);
	const headerLabelClass = joinClasses(
		CARD_LABEL_CLASS,
		'text-amber-600 dark:text-amber-300',
	);
	const headerRowClass = 'flex items-start justify-between gap-4';
	const actionBadgeClass = joinClasses(
		'flex h-12 w-12 shrink-0 items-center justify-center rounded-2xl',
		'border border-white/50 bg-white/70 text-3xl shadow-inner',
		'shadow-amber-500/20 dark:border-white/10 dark:bg-slate-900/60',
		'dark:shadow-slate-900/40',
	);
	const resolutionContainerClass = joinClasses(
		'mt-4 rounded-3xl border border-white/50 bg-white/70 p-4',
		'shadow-inner shadow-amber-500/10 ring-1 ring-white/30',
		'backdrop-blur-sm dark:border-white/10 dark:bg-slate-900/60',
		'dark:shadow-slate-900/40 dark:ring-white/10',
	);
	const timelineListClass = 'space-y-3';
	const timelineItemClass = 'relative pl-6';
	const timelineMarkerClass = joinClasses(
		'absolute left-0 top-2 flex h-3 w-3 items-center justify-center',
		'rounded-full bg-amber-500 shadow-md shadow-amber-500/40',
	);
	const timelineTextClass = joinClasses(
		CARD_BODY_TEXT_CLASS,
		'whitespace-pre-wrap leading-relaxed',
	);
	const summaryContainerClass = joinClasses(
		'rounded-2xl border border-amber-500/30 bg-amber-50/80 p-3',
		'shadow-inner shadow-amber-500/20 dark:border-amber-400/30',
		'dark:bg-amber-400/10',
	);
	const summaryTextClass = joinClasses(
		CARD_BODY_TEXT_CLASS,
		'whitespace-pre-line text-amber-900 dark:text-amber-100',
	);

	return (
		<div className={containerClass} data-state="enter">
			<div className="space-y-3">
				<div className={headerRowClass}>
					<div className="space-y-1">
						<div className={headerLabelClass}>Resolution</div>
						<div className={CARD_TITLE_TEXT_CLASS}>{headerTitle}</div>
						{resolution.player ? (
							<div className={CARD_META_TEXT_CLASS}>
								{`Played by ${playerName}`}
							</div>
						) : null}
					</div>
					{actionIcon || actionName ? (
						<div className="flex flex-col items-end gap-2 text-right">
							<div className={actionBadgeClass} aria-hidden="true">
								{actionIcon ?? 'âœ¦'}
							</div>
							{actionName ? (
								<div className={CARD_META_TEXT_CLASS}>{actionName}</div>
							) : null}
						</div>
					) : null}
				</div>
				{summaryItems.length ? (
					<div className={summaryContainerClass}>
						<div className={joinClasses(CARD_LABEL_CLASS, 'text-amber-700')}>
							Highlights
						</div>
						<ul className="mt-2 space-y-1">
							{summaryItems.map((item, index) => (
								<li key={index} className={summaryTextClass}>
									{item}
								</li>
							))}
						</ul>
					</div>
				) : null}
				<div className={resolutionContainerClass}>
					<div className={joinClasses(CARD_LABEL_CLASS, 'text-slate-600')}>
						Resolution steps
					</div>
					<ol className={timelineListClass}>
						{resolution.visibleLines.map((line, index) => (
							<li key={index} className={timelineItemClass}>
								<span className={timelineMarkerClass} />
								<div className={timelineTextClass}>{line}</div>
							</li>
						))}
					</ol>
				</div>
			</div>
			<div className="mt-6 flex justify-end">
				<button
					type="button"
					onClick={onContinue}
					disabled={!resolution.isComplete}
					className={CONTINUE_BUTTON_CLASS}
				>
					Continue
				</button>
			</div>
		</div>
	);
}

export type { ActionResolutionCardProps };
export { ActionResolutionCard };
