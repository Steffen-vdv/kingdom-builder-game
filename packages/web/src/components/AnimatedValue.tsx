import React, { useEffect, useRef, useState } from 'react';
import type { MutableRefObject } from 'react';
import { useAnimate } from '../utils/useAutoAnimate';

interface AnimatedValueProps {
  value: number;
  format?: (value: number) => string;
}

const AnimatedValue: React.FC<AnimatedValueProps> = ({
  value,
  format = String,
}) => {
  const [delta, setDelta] = useState(0);
  const prev = useRef(value);
  const ref: MutableRefObject<HTMLSpanElement | null> = useAnimate();

  useEffect(() => {
    const diff = value - prev.current;
    if (diff !== 0) {
      setDelta(diff);
      const timer = setTimeout(() => setDelta(0), 800);
      prev.current = value;
      return () => clearTimeout(timer);
    }
  }, [value]);

  const formatted = format(value);
  const diffFormatted = format(Math.abs(delta));

  return (
    <span ref={ref} className="relative inline-block">
      <span key={formatted}>{formatted}</span>
      {delta !== 0 && (
        <span className={delta > 0 ? 'gain-bubble' : 'loss-bubble'}>
          {delta > 0 ? `+${diffFormatted}` : `-${diffFormatted}`}
        </span>
      )}
    </span>
  );
};

export default AnimatedValue;
