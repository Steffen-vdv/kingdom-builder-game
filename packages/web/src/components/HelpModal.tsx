import React from 'react';
import Button from './common/Button';
import { OverviewContent } from '../Overview';

interface HelpModalProps {
  open: boolean;
  onClose: () => void;
}

export default function HelpModal({ open, onClose }: HelpModalProps) {
  if (!open) return null;
  return (
    <div className="fixed inset-0 z-50 flex items-center justify-center bg-black/50 p-4">
      <div className="relative w-full max-w-2xl overflow-y-auto rounded bg-white p-4 text-gray-900 shadow dark:bg-gray-800 dark:text-gray-100 max-h-full">
        <Button
          variant="ghost"
          className="absolute right-2 top-2"
          onClick={onClose}
          aria-label="Close help"
        >
          âœ•
        </Button>
        <OverviewContent />
      </div>
    </div>
  );
}
