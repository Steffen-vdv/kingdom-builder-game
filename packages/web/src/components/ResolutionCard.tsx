import React from 'react';
import type {
	ActionResolution,
	ResolutionSource,
} from '../state/useActionResolution';
import type { TimelineEntry } from './ResolutionTimeline';
import {
	CARD_BASE_CLASS,
	CARD_BODY_TEXT_CLASS,
	CARD_LABEL_CLASS,
	CARD_META_TEXT_CLASS,
	CARD_TITLE_TEXT_CLASS,
	CONTINUE_BUTTON_CLASS,
	NEXT_TURN_BUTTON_CLASS,
	joinClasses,
} from './common/cardStyles';
import { usePlayerAccentClasses } from './common/usePlayerAccentClasses';
import {
	buildTimelineTree,
	buildResolutionTimelineEntries,
	normalizeModifierDescription,
} from './ResolutionTimeline';
import { useSoundEffectsContext } from '../state/SoundEffectsContext';

interface ResolutionLabels {
	title: string;
	player: string;
}

const SOURCE_LABELS: Record<'action' | 'phase', ResolutionLabels> = {
	action: {
		title: 'Action',
		player: 'Played by',
	},
	phase: {
		title: 'Phase',
		player: 'Phase owner',
	},
};

const LEADING_EMOJI_PATTERN =
	/^(?:\p{Extended_Pictographic}(?:\uFE0F|\uFE0E)?(?:\u200D\p{Extended_Pictographic}(?:\uFE0F|\uFE0E)?)*)/u;

function extractLeadingIcon(value: string | undefined) {
	if (!value) {
		return undefined;
	}
	const trimmed = value.trimStart();
	if (!trimmed) {
		return undefined;
	}
	const match = trimmed.match(LEADING_EMOJI_PATTERN);
	if (!match) {
		return undefined;
	}
	const icon = match[0]?.trim();
	return icon && /\p{Extended_Pictographic}/u.test(icon) ? icon : undefined;
}

function isSourceDetail(
	source: ResolutionSource | undefined,
): source is Exclude<ResolutionSource, 'action' | 'phase'> {
	return typeof source === 'object' && source !== null && 'kind' in source;
}

interface ResolutionCardProps {
	title?: string;
	resolution: ActionResolution;
	onContinue: () => void;
	continueMode?: 'continue' | 'next-turn';
}

function resolveSourceLabels(source: ResolutionSource | undefined) {
	if (!source) {
		return SOURCE_LABELS.action;
	}
	if (typeof source === 'string') {
		return SOURCE_LABELS[source] ?? SOURCE_LABELS.action;
	}
	const fallback = SOURCE_LABELS[source.kind] ?? SOURCE_LABELS.action;
	const title = source.label?.trim() || fallback.title;
	return {
		title,
		player: fallback.player,
	};
}

function ResolutionCard({
	title,
	resolution,
	onContinue,
	continueMode = 'continue',
}: ResolutionCardProps) {
	const playerLabel = resolution.player?.name ?? resolution.player?.id ?? null;
	const playerName = playerLabel ?? 'Unknown player';
	const resolveAccentClasses = usePlayerAccentClasses();
	const accentClasses = resolveAccentClasses(resolution.player?.id ?? null);
	const { playUiClick } = useSoundEffectsContext();
	const handleContinueClick = React.useCallback(() => {
		playUiClick();
		onContinue();
	}, [onContinue, playUiClick]);
	const containerClass = joinClasses(
		CARD_BASE_CLASS,
		'pointer-events-auto',
		accentClasses.card,
	);
	const leadingLine = resolution.lines[0]?.trim() ?? '';

	const fallbackActionName = leadingLine
		.replace(/^[\s\p{Extended_Pictographic}\uFE0F\p{Pd}\p{Po}\p{So}]+/u, '')
		.replace(/^Played\s+/u, '')
		.replace(/[\p{Extended_Pictographic}\uFE0F]/gu, '')
		.replace(/\s{2,}/g, ' ')
		.trim();
	const sourceName = isSourceDetail(resolution.source)
		? (resolution.source.name?.trim() ?? '')
		: '';
	const sourceLabel = isSourceDetail(resolution.source)
		? resolution.source.label
		: undefined;
	const rawActionName = (resolution.action?.name ?? '').trim() || sourceName;
	const actionName = rawActionName || fallbackActionName;
	const resolvedLabels = resolveSourceLabels(resolution.source);
	const resolvedSourceKind = isSourceDetail(resolution.source)
		? resolution.source.kind
		: typeof resolution.source === 'string'
			? resolution.source
			: undefined;
	const actorLabel = (resolution.actorLabel ?? '').trim();
	const normalizedActorLabel = actorLabel ? actorLabel.toLocaleLowerCase() : '';
	const normalizedPlayerLabel = resolvedLabels.player
		.trim()
		.toLocaleLowerCase();
	const actorHeaderSubject = actionName
		? !actorLabel || normalizedActorLabel === normalizedPlayerLabel
			? actionName
			: actorLabel
		: actorLabel || actionName;
	const sourceIcon = isSourceDetail(resolution.source)
		? resolution.source.icon?.trim() || extractLeadingIcon(sourceLabel)
		: undefined;
	const fallbackIcon = extractLeadingIcon(leadingLine);
	const actionIcon =
		resolution.action?.icon?.trim() || sourceIcon || fallbackIcon;
	const resolutionLabelBase =
		resolvedSourceKind === 'phase'
			? SOURCE_LABELS.phase.title
			: resolvedLabels.title;
	const resolutionLabel = `${resolutionLabelBase} Resolution`;
	const defaultTitle = title ?? resolutionLabel;
	let headerTitle = actorHeaderSubject || defaultTitle;
	if (resolvedSourceKind === 'phase') {
		const sanitizedPhaseSubject = (actorHeaderSubject || '')
			.replace(LEADING_EMOJI_PATTERN, '')
			.replace(/\s{2,}/g, ' ')
			.trim();
		headerTitle = sanitizedPhaseSubject || defaultTitle;
	}
	const headerLabelClass = joinClasses(
		CARD_LABEL_CLASS,
		'text-amber-600 dark:text-amber-300',
	);
	const headerRowClass = 'flex items-start gap-4';
	const actionBadgeClass = joinClasses(
		'flex h-12 w-12 shrink-0 items-center justify-center rounded-2xl',
		'border border-white/50 bg-white/70 text-3xl shadow-inner',
		'shadow-amber-500/20 dark:border-white/10 dark:bg-slate-900/60',
		'dark:shadow-slate-900/40',
	);
	const resolutionContainerClass = joinClasses(
		'mt-4 rounded-3xl border border-white/50 bg-white/70 p-4',
		'shadow-inner shadow-amber-500/10 ring-1 ring-white/30',
		'backdrop-blur-sm dark:border-white/10 dark:bg-slate-900/60',
		'dark:shadow-slate-900/40 dark:ring-white/10',
	);
	const timelineListClass = 'relative flex flex-col gap-3 pl-4';
	const timelineRailClass = joinClasses(
		'pointer-events-none absolute left-[0.875rem] top-4 bottom-4 w-px',
		'bg-white/30 dark:bg-white/10',
	);
	const timelineTextClass = joinClasses(
		CARD_BODY_TEXT_CLASS,
		'whitespace-pre-wrap leading-relaxed',
	);
	const nestedTimelineTextClass = joinClasses(
		timelineTextClass,
		'text-slate-600 dark:text-slate-300',
	);
	const timelineItemClass = 'relative flex items-start gap-3';
	const primaryMarkerClass = joinClasses(
		'mt-1.5 flex h-3 w-3 shrink-0 items-center justify-center rounded-full',
		'bg-amber-500 shadow-[0_0_0_4px_rgba(251,191,36,0.25)]',
		'dark:bg-amber-400 dark:shadow-[0_0_0_4px_rgba(251,191,36,0.2)]',
	);
	const nestedMarkerClass = joinClasses(
		'mt-1.5 flex h-2.5 w-2.5 shrink-0 items-center justify-center rounded-full',
		'bg-slate-400/80 shadow-[0_0_0_4px_rgba(148,163,184,0.2)]',
		'dark:bg-slate-500 dark:shadow-[0_0_0_4px_rgba(15,23,42,0.45)]',
	);
	const structuredTimeline = React.useMemo(
		() => buildTimelineTree(resolution.visibleTimeline),
		[resolution.visibleTimeline],
	);
	const timelineEntries = React.useMemo(() => {
		const options: Parameters<typeof buildResolutionTimelineEntries>[1] = {};

		if (actionName) {
			options.actionName = actionName;
		}

		if (actionIcon) {
			options.actionIcon = actionIcon;
		}

		return buildResolutionTimelineEntries(structuredTimeline, options);
	}, [actionIcon, actionName, structuredTimeline]);

	const fallbackLines = React.useMemo(() => {
		if (resolution.visibleTimeline.length > 0) {
			return [] as { depth: number; text: string }[];
		}

		function parseLine(line: string) {
			const patterns = [
				/^(?: {3})/,
				/^(?:[ \t]*[•▪‣◦●]\s+)/u,
				/^(?:[ \t]*[↳➜➤➣]\s+)/u,
			];

			let remaining = line;
			let depth = 0;
			let matched = true;

			while (matched) {
				matched = false;
				for (const pattern of patterns) {
					const match = remaining.match(pattern);
					if (match && match[0].length > 0) {
						remaining = remaining.slice(match[0].length);
						depth += 1;
						matched = true;
						break;
					}
				}
			}

			const text = normalizeModifierDescription(remaining.trimStart());
			return { depth, text };
		}

		return resolution.visibleLines.map((line) => parseLine(line));
	}, [resolution.visibleLines, resolution.visibleTimeline]);
	const hasStructuredTimeline = timelineEntries.length > 0;

	function renderEntry(entry: TimelineEntry): React.ReactNode {
		const isSectionRoot = entry.kind === 'section';
		const markerClass = isSectionRoot ? primaryMarkerClass : nestedMarkerClass;
		const itemIndent = !isSectionRoot
			? { marginLeft: `${entry.level * 0.875}rem` }
			: undefined;
		const textClass = isSectionRoot
			? timelineTextClass
			: nestedTimelineTextClass;

		return (
			<div key={entry.key} className={timelineItemClass} style={itemIndent}>
				<span className={markerClass} aria-hidden="true" />
				<div className={textClass}>{entry.text}</div>
			</div>
		);
	}
	const shouldShowContinue = resolution.requireAcknowledgement;
	const isNextTurnMode = continueMode === 'next-turn';
	const continueButtonClass = isNextTurnMode
		? NEXT_TURN_BUTTON_CLASS
		: CONTINUE_BUTTON_CLASS;
	const continueButtonLabel = isNextTurnMode ? 'Next Turn' : 'Continue';
	const continueButtonIcon = isNextTurnMode ? '»' : '→';

	return (
		<div className={containerClass} data-state="enter">
			<div className="space-y-3">
				<div className={headerRowClass}>
					{actionIcon || actionName ? (
						<div className={actionBadgeClass} aria-hidden="true">
							{actionIcon ?? '✦'}
						</div>
					) : null}
					<div className="flex flex-1 items-start justify-between gap-4">
						<div className="space-y-1">
							<div className={headerLabelClass}>{resolutionLabel}</div>
							<div className={CARD_TITLE_TEXT_CLASS}>{headerTitle}</div>
						</div>
						{resolution.player ? (
							<div
								className={joinClasses('text-right', CARD_META_TEXT_CLASS)}
								aria-label="Player"
							>
								{playerName}
							</div>
						) : null}
					</div>
				</div>
				<div className={resolutionContainerClass}>
					<div className={joinClasses(CARD_LABEL_CLASS, 'text-slate-600')}>
						Resolution steps
					</div>
					{hasStructuredTimeline ? (
						<div className={joinClasses(timelineListClass, 'mt-3')}>
							<div aria-hidden="true" className={timelineRailClass} />
							{timelineEntries.map((entry) => renderEntry(entry))}
						</div>
					) : (
						<div className={joinClasses(timelineListClass, 'mt-3')}>
							<div aria-hidden="true" className={timelineRailClass} />
							{fallbackLines.map(({ text, depth }, index) => {
								const markerClass =
									depth === 0 ? primaryMarkerClass : nestedMarkerClass;
								const itemIndent =
									depth > 0
										? {
												marginLeft: `${depth * 0.875}rem`,
											}
										: undefined;
								const textClass =
									depth === 0 ? timelineTextClass : nestedTimelineTextClass;

								return (
									<div
										key={index}
										className={timelineItemClass}
										style={itemIndent}
									>
										<span className={markerClass} aria-hidden="true" />
										<div className={textClass}>{text}</div>
									</div>
								);
							})}
						</div>
					)}
				</div>
			</div>
			{shouldShowContinue ? (
				<div className="mt-6 flex justify-end">
					<button
						type="button"
						onClick={handleContinueClick}
						disabled={!resolution.isComplete}
						className={continueButtonClass}
					>
						<span>{continueButtonLabel}</span>
						<span aria-hidden="true">{continueButtonIcon}</span>
					</button>
				</div>
			) : null}
		</div>
	);
}

export type { ResolutionCardProps };
export { ResolutionCard };
