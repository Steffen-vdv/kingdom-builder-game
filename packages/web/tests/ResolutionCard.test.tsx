/** @vitest-environment jsdom */
import { describe, it, expect, afterEach } from 'vitest';
import { cleanup, render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import React from 'react';
import { ResolutionCard } from '../src/components/ResolutionCard';
import type { ActionResolution } from '../src/state/useActionResolution';

function createResolution(
	overrides: Partial<ActionResolution>,
): ActionResolution {
	return {
		lines: ['First line'],
		visibleLines: ['First line'],
		visibleTimeline: [],
		isComplete: true,
		summaries: [],
		source: 'action',
		requireAcknowledgement: true,
		...overrides,
	} as ActionResolution;
}

describe('<ResolutionCard />', () => {
	afterEach(() => {
		cleanup();
	});
	it('shows labels for action-based resolutions', () => {
		const resolution = createResolution({
			action: {
				id: 'action-id',
				name: 'Test Action',
				icon: '‚öîÔ∏è',
			},
			source: {
				kind: 'action',
				label: 'Action',
				id: 'action-id',
				name: 'Test Action',
				icon: '‚öîÔ∏è',
			},
			actorLabel: 'Played by',
			player: {
				id: 'player-1',
				name: 'Player One',
			},
		});

		render(
			<ResolutionCard
				resolution={resolution}
				title="Resolution Title"
				onContinue={() => {}}
			/>,
		);

		expect(screen.getByText('Action - Test Action')).toBeInTheDocument();
		const actionPlayerLabel = screen.getByLabelText('Player');
		expect(actionPlayerLabel).toHaveTextContent('Player One');
	});

	it('renders custom source metadata when provided', () => {
		const resolution = createResolution({
			source: {
				kind: 'phase',
				label: 'Growth Phase',
				name: 'Growth Phase',
			},
			actorLabel: 'Growth Phase',
			player: {
				id: 'player-2',
				name: 'Player Two',
			},
		});

		render(<ResolutionCard resolution={resolution} onContinue={() => {}} />);

		expect(screen.getByText('Growth Phase')).toBeInTheDocument();
		const phasePlayerLabel = screen.getByLabelText('Player');
		expect(phasePlayerLabel).toHaveTextContent('Player Two');
	});

	it('hides the continue button when acknowledgement is not required', () => {
		const resolution = createResolution({
			requireAcknowledgement: false,
		});
		expect(resolution.requireAcknowledgement).toBe(false);

		const { queryByRole } = render(
			<ResolutionCard resolution={resolution} onContinue={() => {}} />,
		);

		expect(queryByRole('button', { name: 'Continue' })).toBeNull();
	});

	it('separates cost entries from effect entries', () => {
		const resolution = createResolution({
			visibleTimeline: [
				{ text: 'üèóÔ∏è Develop', depth: 0, kind: 'headline' },
				{ text: 'üí≤ Action cost', depth: 1, kind: 'cost' },
				{ text: 'Gold -3', depth: 2, kind: 'cost-detail' },
				{ text: 'ü™Ñ Effect happens', depth: 1, kind: 'change' },
			],
			visibleLines: [],
		});

		render(<ResolutionCard resolution={resolution} onContinue={() => {}} />);

		const costLabel = screen.getByText('Cost');
		const costList = costLabel.nextElementSibling;
		const effectLabel = screen.getByText('Effects');
		const effectList = effectLabel.parentElement?.nextElementSibling ?? null;

		expect(costList).not.toBeNull();
		expect(effectList).not.toBeNull();

		if (!costList || !effectList) {
			throw new Error('Expected cost and effect sections to be rendered');
		}

		expect(within(costList).getByText('üí≤ Action cost')).toBeInTheDocument();
		expect(within(costList).getByText('Gold -3')).toBeInTheDocument();
		expect(within(effectList).queryByText('üí≤ Action cost')).toBeNull();
		expect(
			within(effectList).getByText('ü™Ñ Effect happens'),
		).toBeInTheDocument();
	});

	it('renders grouped cost and effect entries with nested structure', () => {
		const resolution = createResolution({
			action: {
				id: 'action-with-costs',
				name: 'Channel Arcana',
				icon: 'üßô',
			},
			visibleTimeline: [
				{ text: 'üßô Channel Arcana', depth: 0, kind: 'headline' },
				{ text: 'üí≤ Action cost', depth: 1, kind: 'cost' },
				{ text: 'Pay 3 Gold', depth: 2, kind: 'cost-detail' },
				{ text: 'Spend 1 Influence', depth: 3, kind: 'cost-detail' },
				{ text: 'ü™Ñ Conjured effects', depth: 1, kind: 'group' },
				{ text: 'Gain 2 Gold', depth: 2, kind: 'change' },
				{ text: 'Trigger a follow-up', depth: 3, kind: 'change' },
				{ text: 'Grant an enchantment', depth: 2, kind: 'effect' },
			],
			visibleLines: [],
		});

		render(<ResolutionCard resolution={resolution} onContinue={() => {}} />);

		const costHeader = screen.getByText('Cost');
		const costList = costHeader.nextElementSibling as HTMLElement | null;
		const effectsHeader = screen.getByText('Effects');
		const effectsList = effectsHeader.parentElement
			?.nextElementSibling as HTMLElement | null;

		if (!costList) {
			throw new Error('Expected the cost list to be rendered');
		}
		if (!effectsList) {
			throw new Error('Expected the effects list to be rendered');
		}

		expect(within(costList).getByText('üí≤ Action cost')).toBeInTheDocument();
		const primaryCostDetail = within(costList).getByText('Pay 3 Gold');
		const primaryCostWrapper = primaryCostDetail.closest('div[style]');
		expect(primaryCostWrapper).toHaveStyle('margin-left: 0.875rem');

		const nestedCostDetail = within(costList).getByText('Spend 1 Influence');
		const nestedCostWrapper = nestedCostDetail.closest('div[style]');
		expect(nestedCostWrapper).toHaveStyle('margin-left: 1.75rem');

		const effectGroup = within(effectsList).getByText('ü™Ñ Conjured effects');
		const effectGroupWrapper = effectGroup.closest('div[style]');
		expect(effectGroupWrapper).toHaveStyle('margin-left: 0.875rem');

		const nestedEffect = within(effectsList).getByText('Trigger a follow-up');
		const nestedEffectWrapper = nestedEffect.closest('div[style]');
		expect(nestedEffectWrapper).toHaveStyle('margin-left: 2.625rem');

		const siblingEffect = within(effectsList).getByText('Grant an enchantment');
		const siblingEffectWrapper = siblingEffect.closest('div[style]');
		expect(siblingEffectWrapper).toHaveStyle('margin-left: 1.75rem');
	});

	it('renders fallback layout for simple phase-style lines', () => {
		const resolution = createResolution({
			source: 'phase',
			actorLabel: 'Growth Phase',
			visibleTimeline: [],
			visibleLines: [
				'Growth Phase summary',
				'   Rewards trigger',
				'   ‚Ä¢ Gain 2 Gold',
				'   ‚Ä¢ Gain 1 Happiness',
				'‚Ü≥ Aftermath resolves',
			],
		});

		const { container } = render(
			<ResolutionCard resolution={resolution} onContinue={() => {}} />,
		);

		expect(screen.queryByText('Cost')).toBeNull();
		expect(screen.queryByText('Effects')).toBeNull();

		const stepsLabel = screen.getByText('Resolution steps');
		const fallbackList = stepsLabel.nextElementSibling as HTMLElement | null;

		if (!fallbackList) {
			throw new Error('Expected fallback resolution list to be rendered');
		}

		expect(
			within(fallbackList).getByText('Growth Phase summary'),
		).toBeInTheDocument();
		const rewardLine = within(fallbackList).getByText('Rewards trigger');
		expect(rewardLine.closest('div[style]')).toHaveStyle(
			'margin-left: 0.875rem',
		);

		const gainGoldLine = within(fallbackList).getByText('Gain 2 Gold');
		expect(gainGoldLine.closest('div[style]')).toHaveStyle(
			'margin-left: 1.75rem',
		);

		const aftermathLine = within(fallbackList).getByText('Aftermath resolves');
		expect(aftermathLine.closest('div[style]')).toHaveStyle(
			'margin-left: 0.875rem',
		);

		expect(container).toMatchSnapshot();
	});
});
