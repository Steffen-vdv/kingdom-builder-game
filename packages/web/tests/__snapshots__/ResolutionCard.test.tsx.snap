// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ResolutionCard /> > renders fallback layout for simple phase-style lines 1`] = `
<div>
  <div
    class="hover-card-transition w-full rounded-3xl border border-white/60 bg-white/80 p-6 shadow-2xl shadow-amber-500/10 dark:border-white/10 dark:bg-slate-900/80 dark:shadow-slate-900/60 frosted-surface pointer-events-auto"
    data-state="enter"
  >
    <div
      class="space-y-3"
    >
      <div
        class="flex items-start gap-4"
      >
        <div
          aria-hidden="true"
          class="flex h-12 w-12 shrink-0 items-center justify-center rounded-2xl border border-white/50 bg-white/70 text-3xl shadow-inner shadow-amber-500/20 dark:border-white/10 dark:bg-slate-900/60 dark:shadow-slate-900/40"
        >
          âœ¦
        </div>
        <div
          class="flex flex-1 items-start justify-between gap-4"
        >
          <div
            class="space-y-1"
          >
            <div
              class="text-xs font-semibold uppercase tracking-[0.3em] text-slate-500 dark:text-slate-300 text-amber-600 dark:text-amber-300"
            >
              Resolution
            </div>
            <div
              class="text-lg font-semibold tracking-tight text-slate-900 dark:text-slate-100"
            >
              Phase - Growth Phase
            </div>
          </div>
        </div>
      </div>
      <div
        class="mt-4 rounded-3xl border border-white/50 bg-white/70 p-4 shadow-inner shadow-amber-500/10 ring-1 ring-white/30 backdrop-blur-sm dark:border-white/10 dark:bg-slate-900/60 dark:shadow-slate-900/40 dark:ring-white/10"
      >
        <div
          class="text-xs font-semibold uppercase tracking-[0.3em] text-slate-500 dark:text-slate-300 text-slate-600"
        >
          Resolution steps
        </div>
        <div
          class="relative flex flex-col gap-3 pl-4 mt-3"
        >
          <div
            aria-hidden="true"
            class="pointer-events-none absolute left-[0.875rem] top-4 bottom-4 w-px bg-white/30 dark:bg-white/10"
          />
          <div
            class="relative flex items-start gap-3"
          >
            <span
              aria-hidden="true"
              class="mt-1.5 flex h-3 w-3 shrink-0 items-center justify-center rounded-full bg-amber-500 shadow-[0_0_0_4px_rgba(251,191,36,0.25)] dark:bg-amber-400 dark:shadow-[0_0_0_4px_rgba(251,191,36,0.2)]"
            />
            <div
              class="text-sm text-slate-700 dark:text-slate-200 whitespace-pre-wrap leading-relaxed"
            >
              Growth Phase summary
            </div>
          </div>
          <div
            class="relative flex items-start gap-3"
            style="margin-left: 0.875rem;"
          >
            <span
              aria-hidden="true"
              class="mt-1.5 flex h-2.5 w-2.5 shrink-0 items-center justify-center rounded-full bg-slate-400/80 shadow-[0_0_0_4px_rgba(148,163,184,0.2)] dark:bg-slate-500 dark:shadow-[0_0_0_4px_rgba(15,23,42,0.45)]"
            />
            <div
              class="text-sm text-slate-700 dark:text-slate-200 whitespace-pre-wrap leading-relaxed text-slate-600 dark:text-slate-300"
            >
              Rewards trigger
            </div>
          </div>
          <div
            class="relative flex items-start gap-3"
            style="margin-left: 1.75rem;"
          >
            <span
              aria-hidden="true"
              class="mt-1.5 flex h-2.5 w-2.5 shrink-0 items-center justify-center rounded-full bg-slate-400/80 shadow-[0_0_0_4px_rgba(148,163,184,0.2)] dark:bg-slate-500 dark:shadow-[0_0_0_4px_rgba(15,23,42,0.45)]"
            />
            <div
              class="text-sm text-slate-700 dark:text-slate-200 whitespace-pre-wrap leading-relaxed text-slate-600 dark:text-slate-300"
            >
              Gain 2 Gold
            </div>
          </div>
          <div
            class="relative flex items-start gap-3"
            style="margin-left: 1.75rem;"
          >
            <span
              aria-hidden="true"
              class="mt-1.5 flex h-2.5 w-2.5 shrink-0 items-center justify-center rounded-full bg-slate-400/80 shadow-[0_0_0_4px_rgba(148,163,184,0.2)] dark:bg-slate-500 dark:shadow-[0_0_0_4px_rgba(15,23,42,0.45)]"
            />
            <div
              class="text-sm text-slate-700 dark:text-slate-200 whitespace-pre-wrap leading-relaxed text-slate-600 dark:text-slate-300"
            >
              Gain 1 Happiness
            </div>
          </div>
          <div
            class="relative flex items-start gap-3"
            style="margin-left: 0.875rem;"
          >
            <span
              aria-hidden="true"
              class="mt-1.5 flex h-2.5 w-2.5 shrink-0 items-center justify-center rounded-full bg-slate-400/80 shadow-[0_0_0_4px_rgba(148,163,184,0.2)] dark:bg-slate-500 dark:shadow-[0_0_0_4px_rgba(15,23,42,0.45)]"
            />
            <div
              class="text-sm text-slate-700 dark:text-slate-200 whitespace-pre-wrap leading-relaxed text-slate-600 dark:text-slate-300"
            >
              Aftermath resolves
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mt-6 flex justify-end"
    >
      <button
        class="inline-flex items-center justify-center rounded-2xl bg-amber-500 cursor-pointer px-4 py-2 text-sm font-semibold text-white shadow-lg shadow-amber-500/40 transition hover:bg-amber-400 focus:outline-none focus-visible:ring focus-visible:ring-amber-500/60 disabled:cursor-not-allowed disabled:bg-amber-500/60 disabled:shadow-none"
        type="button"
      >
        Continue
      </button>
    </div>
  </div>
</div>
`;
